compiler/rustc_mir_build/src/build/matches/mod.rs

```diff
@@ -1903,11 +1910,30 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         // Extract the match-pair from the highest priority candidate and build a test from it.
         let (match_place, test) = self.pick_test(candidates);
 
+        // Patterns like `Some(A) | Some(B)` need spans for their target branches separately.
+        // Keep it in original order so that our condition id assignment can be more intuitive.
+        let mut mcdc_coverage_targets: Option<Vec<_>> =
+            self.tcx.sess.instrument_coverage_mcdc().then(Vec::new);
+
+        let catch_full_matched =
+            |candidate: &Candidate<'_, 'tcx>, match_pair: &MatchPair<'_, 'tcx>| {
+                if let Some(targets) = mcdc_coverage_targets.as_mut() {
+                    targets.push((candidate.extra_data.span, (match_pair.pattern.span, None)));
+                }
+            };
+
         // For each of the N possible test outcomes, build the vector of candidates that applies if
         // the test has that particular outcome.
         let (remaining_candidates, target_candidates) =
-            self.sort_candidates(match_place, &test, candidates);
-
+            self.sort_candidates(match_place, &test, candidates, catch_full_matched);
+
+        let mut add_mcdc_coverage_matched_blk = |pattern_span: Span, blk: BasicBlock| {
+            let Some(targets) = mcdc_coverage_targets.as_mut() else { return };
+            *targets
+                .iter_mut()
+                .find_map(|(span, (_, blk_opt))| (*span == pattern_span).then_some(blk_opt))
+                .expect("target_candidates only are part of candidates") = Some(blk);
+        };
```

Comment:
You won't need to use span equality to recover the data you need anymore. You can remove this.

Reply:
Done. Not understood it well until you point out. Now it looks more comprehensive and simple!